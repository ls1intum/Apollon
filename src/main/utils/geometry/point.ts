export interface IPoint {
  x: number;
  y: number;
}

export class Point implements IPoint {
  constructor(
    public x: number = 0,
    public y: number = 0,
  ) {}

  get length(): number {
    return Math.sqrt(this.x * this.x + this.y * this.y);
  }

  normalize(): Point {
    return new Point(this.x / this.length, this.y / this.length);
  }

  add(x: number, y: number): Point;
  add(x: Point): Point;
  add(x: number | Point, y?: number): Point {
    if (x instanceof Point) {
      return new Point(this.x + x.x, this.y + x.y);
    }
    if (typeof y === 'number') {
      return new Point(this.x + x, this.y + y);
    }
    return this.clone();
  }

  subtract(x: number, y: number): Point;
  subtract(x: Point): Point;
  subtract(x: number | Point, y?: number): Point {
    if (x instanceof Point) {
      return new Point(this.x - x.x, this.y - x.y);
    }
    if (typeof y === 'number') {
      return new Point(this.x - x, this.y - y);
    }
    return this.clone();
  }

  round(radix: number = 10): Point {
    return new Point(Math.round(this.x / radix) * radix, Math.round(this.y / radix) * radix);
  }

  scale(factor: number) {
    return new Point(this.x * factor, this.y * factor);
  }

  clone() {
    return new Point(this.x, this.y);
  }
}
