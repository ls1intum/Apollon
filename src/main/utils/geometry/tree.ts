import { UMLContainer } from '../../services/uml-container/uml-container';
import { IUMLElement, UMLElement } from '../../services/uml-element/uml-element';
import { UMLElementState } from '../../services/uml-element/uml-element-types';

export function filterRoots(ids: string[], elements: UMLElementState): string[] {
  const getSelection = (root: IUMLElement): string[] => {
    if (ids.includes(root.id)) return [root.id];

    if (UMLContainer.isUMLContainer(root)) {
      return root.ownedElements.reduce<string[]>((selection, id) => [...selection, ...getSelection(elements[id])], []);
    }
    return [];
  };

  return Object.values(elements)
    .filter((element) => !element.owner)
    .reduce<string[]>((selection, element) => [...selection, ...getSelection(element)], []);
}

/**
 * returns the ids of all elements in the hierarchy above the element with the specified id
 * @param id element id which parents should be found
 * @param elements elements in state
 */
export function getAllParents(id: string, elements: UMLElementState): string[] {
  const getParents = (element: IUMLElement): string[] => {
    // reached top
    if (element.owner === null) return [];
    return [element.owner, ...getParents(elements[element.owner])];
  };
  return getParents(elements[id]);
}

export function getChildren(ids: string[], elements: UMLElementState): string[] {
  const result: string[] = [];

  for (const id of ids) {
    const owner = elements[id];
    if (!owner) continue;

    if (UMLContainer.isUMLContainer(owner)) {
      result.push(...getChildren(owner.ownedElements, elements));
    }
    result.push(owner.id);
  }

  return result;
}

export function clone(element: UMLElement, elements: UMLElement[]): UMLElement[] {
  if (!UMLContainer.isUMLContainer(element)) {
    return [element.clone()];
  }

  const result: UMLElement[] = [];
  const cloned = element.clone<UMLContainer>();
  const { ownedElements } = element;
  for (const id of ownedElements) {
    const child = elements.find((prev) => prev.id === id);
    if (!child) {
      continue;
    }

    const [clonedChild, ...clonedChildren] = clone(child, elements);
    clonedChild.owner = cloned.id;

    const index = cloned.ownedElements.findIndex((x) => x === id);
    cloned.ownedElements[index] = clonedChild.id;
    result.push(clonedChild, ...clonedChildren);
  }

  return [cloned, ...result];
}
