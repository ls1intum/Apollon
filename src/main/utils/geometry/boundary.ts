import { Point } from './point';

export interface IBoundary {
  x: number;
  y: number;
  width: number;
  height: number;
}

export function computeBoundingBox(points: Point[]): IBoundary {
  if (points.length === 0) {
    return { x: 0, y: 0, width: 0, height: 0 };
  }

  const firstPoint = points[0];

  let minX = firstPoint.x;
  let minY = firstPoint.y;

  let maxX = firstPoint.x;
  let maxY = firstPoint.y;

  for (const p of points) {
    if (p.x < minX) minX = p.x;
    if (p.y < minY) minY = p.y;

    if (p.x > maxX) maxX = p.x;
    if (p.y > maxY) maxY = p.y;
  }

  return {
    x: minX,
    y: minY,
    width: maxX - minX,
    height: maxY - minY,
  };
}

export function computeBoundingBoxForElements(elements: { bounds: IBoundary }[]): IBoundary {
  if (!elements.length) {
    return { x: 0, y: 0, width: 0, height: 0 };
  }
  const boundaries: IBoundary[] = elements.map<IBoundary>((element) => ({ ...element.bounds }));
  const x = Math.min(...boundaries.map((bounds) => bounds.x));
  const y = Math.min(...boundaries.map((bounds) => bounds.y));
  const width = Math.max(...boundaries.map((bounds) => bounds.x + bounds.width)) - x;
  const height = Math.max(...boundaries.map((bounds) => bounds.y + bounds.height)) - y;
  return { x, y, width, height };
}

export function computeDimension(scale: number, value: number, isCircle?: boolean): number {
  if (isCircle && scale === 1) {
    return value * scale;
  } else {
    return Math.round((value * scale) / 10) * 10;
  }
}
