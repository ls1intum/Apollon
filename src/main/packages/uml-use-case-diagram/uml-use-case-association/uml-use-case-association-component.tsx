import React, { SFC } from 'react';
import { Text } from '../../../components/controls/text/text';
import { Point } from '../../../utils/geometry/point';
import { UMLUseCaseAssociation } from './uml-use-case-association';

export const UMLUseCaseAssociationComponent: SFC<Props> = ({ element }) => {
  const [start, end] = element.path.map((p) => new Point(p.x, p.y));
  const line = end.subtract(start);
  const norm = line.normalize();
  const center = start.add(norm.scale(0.5 * line.length));

  return (
    <g>
      <path
        id={`textpath-${element.id}`}
        d={`
        M ${start.x} ${start.y - 10}
        L ${end.x} ${end.y - 10}
    `}
      />
      <Text
        noX
        noY
        fill={element.textColor}
        transform={
          norm.x < 0
            ? `
            translate(${center.x}, ${center.y})
            rotate(180)
            translate(${-center.x}, ${-center.y})
          `
            : undefined
        }
      >
        <textPath xlinkHref={`#textpath-${element.id}`} startOffset="50%">
          {element.name}
        </textPath>
      </Text>
      <polyline
        points={element.path.map((point) => `${point.x} ${point.y}`).join(',')}
        stroke={element.strokeColor || 'black'}
        fill="none"
        strokeWidth={1}
      />
    </g>
  );
};

interface Props {
  element: UMLUseCaseAssociation;
}
