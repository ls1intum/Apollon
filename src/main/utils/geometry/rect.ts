import { Direction } from '../../services/uml-element/uml-element-port';
import { IBoundary } from './boundary';
import { Point } from './point';

export function position(bounds: IBoundary, direction: Direction): { point: Point; offset: Point } {
  const { x, y, width, height } = bounds;
  const offset = 40;
  switch (direction) {
    case Direction.Up:
      return {
        point: new Point(x + width, y),
        offset: new Point(x + width, y - offset),
      };
    case Direction.Right:
      return {
        point: new Point(x + width, y + height),
        offset: new Point(x + width + offset, y + height),
      };
    case Direction.Down:
      return {
        point: new Point(x + width, y + height),
        offset: new Point(x + width, y + height + offset),
      };
    case Direction.Left:
      return {
        point: new Point(x, y + height),
        offset: new Point(x - offset, y + height),
      };
  }
}
