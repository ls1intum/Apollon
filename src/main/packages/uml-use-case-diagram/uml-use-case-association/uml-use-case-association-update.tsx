import React, { Component, ComponentClass } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import styled from 'styled-components';
import { UseCaseRelationshipType } from '..';
import { Button } from '../../../components/controls/button/button';
import { ColorButton } from '../../../components/controls/color-button/color-button';
import { Divider } from '../../../components/controls/divider/divider';
import { Dropdown } from '../../../components/controls/dropdown/dropdown';
import { ExchangeIcon } from '../../../components/controls/icon/exchange';
import { TrashIcon } from '../../../components/controls/icon/trash';
import { Textfield } from '../../../components/controls/textfield/textfield';
import { Header } from '../../../components/controls/typography/typography';
import { I18nContext } from '../../../components/i18n/i18n-context';
import { localized } from '../../../components/i18n/localized';
import { ModelState } from '../../../components/store/model-state';
import { StylePane } from '../../../components/style-pane/style-pane';
import { UMLElementRepository } from '../../../services/uml-element/uml-element-repository';
import { UMLRelationshipRepository } from '../../../services/uml-relationship/uml-relationship-repository';
import { UMLUseCaseExtend } from '../uml-use-case-extend/uml-use-case-extend';
import { UMLUseCaseGeneralization } from '../uml-use-case-generalization/uml-use-case-generalization';
import { UMLUseCaseInclude } from '../uml-use-case-include/uml-use-case-include';
import { UMLUseCaseAssociation } from './uml-use-case-association';

const Flex = styled.div`
  display: flex;
  align-items: baseline;
  justify-content: space-between;
`;

type State = { colorOpen: boolean };

class UseCaseAssociationUpdate extends Component<Props, State> {
  state = { colorOpen: false };

  private toggleColor = () => {
    this.setState((state) => ({
      colorOpen: !state.colorOpen,
    }));
  };

  render() {
    const { element } = this.props;

    return (
      <div>
        <section>
          {element.type === UseCaseRelationshipType.UseCaseAssociation ? (
            <Flex>
              <Textfield value={element.name} placeholder="..." onChange={this.rename(element.id)} autoFocus />
              <ColorButton onClick={this.toggleColor} />
              <Button color="link" tabIndex={-1} onClick={() => this.props.delete(element.id)}>
                <TrashIcon />
              </Button>
            </Flex>
          ) : (
            <Flex>
              <Header gutter={false} style={{ flexGrow: 1 }}>
                {
                  {
                    [UseCaseRelationshipType.UseCaseAssociation]: this.props.translate(
                      'packages.UseCaseDiagram.UseCaseAssociation',
                    ),
                    [UseCaseRelationshipType.UseCaseGeneralization]: this.props.translate(
                      'packages.UseCaseDiagram.UseCaseGeneralization',
                    ),
                    [UseCaseRelationshipType.UseCaseInclude]: this.props.translate(
                      'packages.UseCaseDiagram.UseCaseInclude',
                    ),
                    [UseCaseRelationshipType.UseCaseExtend]: this.props.translate(
                      'packages.UseCaseDiagram.UseCaseExtend',
                    ),
                  }[element.type]
                }
              </Header>
              <ColorButton onClick={this.toggleColor} />
              <Button color="link" tabIndex={-1} onClick={() => this.props.flip(element.id)}>
                <ExchangeIcon />
              </Button>
              <Button color="link" tabIndex={-1} onClick={() => this.props.delete(element.id)}>
                <TrashIcon />
              </Button>
            </Flex>
          )}
        </section>
        <section>
          <Divider />
          <Dropdown value={element.type} onChange={this.onChange}>
            <Dropdown.Item value={UseCaseRelationshipType.UseCaseAssociation}>
              {this.props.translate('packages.UseCaseDiagram.UseCaseAssociation')}
            </Dropdown.Item>
            <Dropdown.Item value={UseCaseRelationshipType.UseCaseGeneralization}>
              {this.props.translate('packages.UseCaseDiagram.UseCaseGeneralization')}
            </Dropdown.Item>
            <Dropdown.Item value={UseCaseRelationshipType.UseCaseInclude}>
              {this.props.translate('packages.UseCaseDiagram.UseCaseInclude')}
            </Dropdown.Item>
            <Dropdown.Item value={UseCaseRelationshipType.UseCaseExtend}>
              {this.props.translate('packages.UseCaseDiagram.UseCaseExtend')}
            </Dropdown.Item>
          </Dropdown>
        </section>
        <StylePane
          open={this.state.colorOpen}
          element={element}
          onColorChange={this.props.update}
          lineColor
          textColor={element.type !== UseCaseRelationshipType.UseCaseGeneralization}
        />
      </div>
    );
  }
  private rename = (id: string) => (value: string) => {
    this.props.update(id, { name: value });
  };

  private onChange = (type: keyof typeof UseCaseRelationshipType) => {
    const { element, update } = this.props;
    update(element.id, { type });
  };
}

type OwnProps = {
  element: UMLUseCaseAssociation | UMLUseCaseGeneralization | UMLUseCaseInclude | UMLUseCaseExtend;
};

type StateProps = {};

type DispatchProps = {
  update: typeof UMLElementRepository.update;
  delete: typeof UMLElementRepository.delete;
  flip: typeof UMLRelationshipRepository.flip;
};

type Props = OwnProps & StateProps & DispatchProps & I18nContext;

const enhance = compose<ComponentClass<OwnProps>>(
  localized,
  connect<StateProps, DispatchProps, OwnProps, ModelState>(null, {
    update: UMLElementRepository.update,
    delete: UMLElementRepository.delete,
    flip: UMLRelationshipRepository.flip,
  }),
);

export const UMLUseCaseAssociationUpdate = enhance(UseCaseAssociationUpdate);
