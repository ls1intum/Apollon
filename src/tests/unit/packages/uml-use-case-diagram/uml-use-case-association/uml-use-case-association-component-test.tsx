import * as React from 'react';
import { wrappedRender } from '../../../test-utils/render';
import { Point } from '../../../../../main/utils/geometry/point';
import { UMLUseCaseAssociationComponent } from '../../../../../main/packages/uml-use-case-diagram/uml-use-case-association/uml-use-case-association-component';
import { UMLUseCaseAssociation } from '../../../../../main/packages/uml-use-case-diagram/uml-use-case-association/uml-use-case-association';

it('render the uml-use-case-association', () => {
  const umlUseCaseAssociation: UMLUseCaseAssociation = new UMLUseCaseAssociation({
    id: 'd37b8ce3-17d2-4432-8fff-6c38ff2a1334',
    name: 'UseCaseAssociation',
    path: [new Point(0, 0), new Point(100, 100)],
  });
  const { getByText, baseElement } = wrappedRender(
    <svg>
      <UMLUseCaseAssociationComponent element={umlUseCaseAssociation} />
    </svg>,
  );
  expect(getByText(umlUseCaseAssociation.name)).toBeInTheDocument();
  expect(baseElement).toMatchSnapshot();
});
