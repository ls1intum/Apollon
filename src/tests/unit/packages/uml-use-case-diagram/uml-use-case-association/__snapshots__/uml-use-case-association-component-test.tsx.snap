// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render the uml-use-case-association 1`] = `
<body>
  <div>
    <svg>
      <g>
        <path
          d="
        M 0 -10
        L 100 90
    "
          id="textpath-d37b8ce3-17d2-4432-8fff-6c38ff2a1334"
        />
        <text
          dominant-baseline="middle"
          dy="20px"
          font-weight="bold"
          pointer-events="none"
          text-anchor="middle"
        >
          <textpath
            startOffset="50%"
            xlink:href="#textpath-d37b8ce3-17d2-4432-8fff-6c38ff2a1334"
          >
            UseCaseAssociation
          </textpath>
        </text>
        <polyline
          class="sc-bczRLJ gQkeMM"
          fill="none"
          points="0 0,100 100"
          stroke="black"
          stroke-width="1"
        />
      </g>
    </svg>
  </div>
</body>
`;
